### DISCUSSION TITLE
Formatting of .npz file is difficult to understand

### DISCUSSION THREAD

**Discussion by ntm529** [geoelements/gns#19] [Q&A] [answered]
I had a question about the formatting of the npz files. I haven't worked with this format, and couldn't find good documentation about it online. It appears to return three tuples of values, each of which being a list (of some kind?) of a 2x1 vector.  Attached is the file, as I read it within the spyder IDE. I also circled the parts that I'm confused about in the other picture. I couldn't find anything in the notes for the meeting that we had. If there's any  explanation, documentation from the lab or website, or just the internet in general, on how to create the format desired, then I would be extremely grateful. I attached the training file that I am trying to "translate" into this npz format as well. 
For reference, in my particle simulation I am trying to input {id, x, y, xvelocity, yvelocity, xacceleration, yaccerleration}
![unnamed](https://user-images.githubusercontent.com/88118608/160919167-e041a39d-2c1e-4de3-8ba5-20bfeb33c1b4.png)
![unnamed](https://user-images.githubusercontent.com/88118608/160919181-0c025be0-5e2c-4f49-9037-4e63ee87bc1b.png)
![image](https://user-images.githubusercontent.com/88118608/160919855-3fe91261-ade4-40a4-82d8-548cf7b8a11e.png)

**Comment by jpvantassel** [comment, marked as answer]
Hi, 

This PyTorch implementation of DeepMind's Graph Network Simulator no longer uses their `.tfrecord` format, instead we use a modified data structure stored in the `.npz` format. For reference `.npz` is a compressed binary file format supported by `numpy` (see https://numpy.org/doc/stable/reference/generated/numpy.savez_compressed.html#numpy.savez_compressed for more details). The data stored inside the `.npz` is organized as a dictionary of tuples, where each tuple (i.e., entry in the dictionary) is for a different training trajectory. Each `tuple` is of the form `(position, particle_type)` where `position` and `particle_type` are numpy ndarrays of `float`s and `int`s respectively. `position` is a 3-D tensor of shape `(n_time_steps, n_particles, n_dimensions)` which track the position `(x, y)` or `(x, y, z)` of the particle in the domain. `particle_type` is a 1-D tensor of shape `(n_particles)` where each entry holds the type of each particle (denoted by `int`s) during the associated trajectory. Note that the Graph Network Simulator does not require the input of velocities or accelerations, only positions.

Here is an example for generating a random trajectory:
```
import numpy as np
rng = np.random.default_rng(seed=1824)
n_time_steps = 10
n_particles = 2
n_dimensions = 2
trajectories = {}
trajectories["trajectory_0"] = (rng.random((n_time_steps, n_particles, n_dimensions), dtype=np.float32),
np.full((n_particles,), 1,dtype=int))
np.savez_compressed("example.npz", **trajectories)
```
And to load that data back in
```
data = np.load("example.npz", allow_pickle=True)
print(len(data["trajectory_0"])) # -> 2
print(data["trajectory_0"][0].shape) # -> (10, 2, 2)
print(data["trajectory_0"][1].shape) # -> (2,)
```
Hope that helps clarify things,
Joe
